File: mseed/tests/test_mseed.py
# SPDX-License-Identifier: MIT
"""
miniSEED writer tests.

Always: round-trip through this package's own minimal reader (structural check).
If ObsPy is installed: also validate against it (the gold standard) — this is the
proof of spec-compliance; run it on a machine with ObsPy.
"""
import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from nic_mseed.steim import STEIM1, STEIM2
from nic_mseed.mseed import write_stream, read_stream

_p = _f = 0
def check(name, cond):
    global _p, _f
    if cond: _p += 1; print(f"  PASS  {name}")
    else:    _f += 1; print(f"  FAIL  {name}")


def main():
    print("miniSEED writer tests")
    import itertools, random
    random.seed(7)
    samples = list(itertools.accumulate([random.randint(-2000, 2000) for _ in range(1500)]))
    start_unix = 1_748_000_000

    for ver, vn in ((STEIM1, "S1"), (STEIM2, "S2")):
        for reclen in (512, 4096):
            blob = write_stream(samples, start_unix=start_unix, start_frac=0.25,
                                sample_rate_hz=100.0, network="NQ", station="QK001",
                                location="", channel="HHZ", version=ver, reclen=reclen)
            check(f"{vn} reclen={reclen}: byte length is whole records",
                  len(blob) % reclen == 0 and len(blob) > 0)
            recs = read_stream(blob)
            recovered = [s for r in recs for s in r["samples"]]
            check(f"{vn} reclen={reclen}: round-trip samples", recovered == samples)
            check(f"{vn} reclen={reclen}: codes preserved",
                  recs[0]["net"] == "NQ" and recs[0]["sta"] == "QK001"
                  and recs[0]["cha"] == "HHZ")
            check(f"{vn} reclen={reclen}: rate preserved", abs(recs[0]["rate_hz"] - 100.0) < 1e-9)
            check(f"{vn} reclen={reclen}: start time preserved",
                  recs[0]["start_unix"] == start_unix and abs(recs[0]["start_frac"] - 0.25) < 1e-4)
            # second record's start advances by (samples_in_rec0 / rate)
            if len(recs) > 1:
                n0 = recs[0]["nsamples"]
                t0 = recs[0]["start_unix"] + recs[0]["start_frac"]
                t1 = recs[1]["start_unix"] + recs[1]["start_frac"]
                check(f"{vn} reclen={reclen}: record-2 time = t0 + n0/rate",
                      abs((t1 - t0) - n0 / 100.0) < 1e-3)

    # sub-Hz rate encodes as a negative factor (period)
    blob = write_stream([1, 2, 3, 4, 5], start_unix=start_unix, sample_rate_hz=0.1,
                        network="NQ", station="SLOW", location="", channel="LHZ")
    check("sub-Hz rate round-trips", abs(read_stream(blob)[0]["rate_hz"] - 0.1) < 1e-9)

    # Optional gold-standard validation against ObsPy.
    try:
        import io
        from obspy import read as obspy_read
        blob = write_stream(samples, start_unix=start_unix, sample_rate_hz=100.0,
                            network="NQ", station="QK001", location="", channel="HHZ",
                            version=STEIM2, reclen=512)
        st = obspy_read(io.BytesIO(blob))
        ok = list(st[0].data.astype(int)) == samples
        check("ObsPy reads our Steim-2 miniSEED and samples match", ok)
        check("ObsPy sees rate 100 Hz", abs(st[0].stats.sampling_rate - 100.0) < 1e-6)
    except ImportError:
        print("  SKIP  ObsPy not installed — run this test on a machine with ObsPy "
              "for gold-standard validation")

    print(f"\nResult: {_p}/{_p+_f} PASS | {_f} FAIL")
    sys.exit(0 if _f == 0 else 1)


if __name__ == "__main__":
    main()

File: mseed/c/test/test_mseed.c
/* SPDX-License-Identifier: MIT
 * miniSEED writer/reader tests: byte-exact vs the Python reference, round-trip,
 * preserved codes/rate/time, and sub-Hz rate encoding. */
#include "nic_mseed.h"
#include "vectors.h"
#include <stdio.h>
#include <string.h>
#include <math.h>

static int passed = 0, failed = 0;
static void check(const char *name, int cond) {
    if (cond) { passed++; printf("  PASS  %s\n", name); }
    else      { failed++; printf("  FAIL  %s\n", name); }
}

#define CAP 8192

int main(void) {
    printf("miniSEED writer tests (C)\n");

    const size_t NB = sizeof(VEC_B) / sizeof(VEC_B[0]);
    uint8_t out[CAP];
    nic_mseed_params_t p = { "NQ", "ST01", "", "HHZ", 100.0, NIC_STEIM2, 512 };

    long n = nic_mseed_write_stream(&p, VEC_B, NB, 1700000000, 0.25, 1, out, CAP);
    check("write_stream returns whole records", n > 0 && n % 512 == 0);
    check("byte-exact vs Python ref (miniSEED stream)",
          n == (long)sizeof(MSEED_S2) && memcmp(out, MSEED_S2, sizeof(MSEED_S2)) == 0);

    /* round-trip: read every record back and compare samples + header */
    int32_t rec_samp[2048];
    size_t got = 0; int ok = 1;
    nic_mseed_rechdr_t h0; int have_h0 = 0;
    for (long off = 0; off < n; off += 512) {
        nic_mseed_rechdr_t h;
        int32_t s[1024];
        if (nic_mseed_read_record(out + off, (size_t)(n - off), &h, s, 1024) != 0) { ok = 0; break; }
        if (!have_h0) { h0 = h; have_h0 = 1; }
        for (uint16_t k = 0; k < h.nsamples && got < 2048; k++) rec_samp[got++] = s[k];
    }
    check("round-trip: all records decode", ok);
    int same = (got == NB);
    for (size_t k = 0; same && k < NB; k++) if (rec_samp[k] != VEC_B[k]) same = 0;
    check("round-trip: samples match", same);
    check("round-trip: codes preserved",
          have_h0 && strcmp(h0.network, "NQ") == 0 && strcmp(h0.station, "ST01") == 0 &&
          strcmp(h0.channel, "HHZ") == 0);
    check("round-trip: rate preserved (100 Hz)", have_h0 && fabs(h0.rate_hz - 100.0) < 1e-9);
    check("round-trip: start time preserved",
          have_h0 && h0.start_unix == 1700000000 && fabs(h0.start_frac - 0.25) < 1e-6);

    /* second record's start = t0 + n0/rate */
    if (n > 512) {
        nic_mseed_rechdr_t h1; int32_t s[1024];
        nic_mseed_read_record(out + 512, (size_t)(n - 512), &h1, s, 1024);
        double t0 = 1700000000 + 0.25;
        double expect = t0 + (double)h0.nsamples / 100.0;
        double actual = (double)h1.start_unix + h1.start_frac;
        check("record-2 time = t0 + n0/rate", fabs(actual - expect) < 1.0 / 10000.0 + 1e-9);
    }

    /* sub-Hz rate encodes as a negative factor (period) and round-trips */
    int32_t few[5] = { 1, 2, 3, 4, 5 };
    nic_mseed_params_t pslow = { "NQ", "ST01", "", "LHZ", 0.1, NIC_STEIM2, 512 };
    long ns = nic_mseed_write_stream(&pslow, few, 5, 1700000000, 0.0, 1, out, CAP);
    int16_t f, m; nic_mseed_rate_factor_mult(0.1, &f, &m);
    check("sub-Hz rate -> negative factor", f == -10 && m == 1);
    if (ns > 0) {
        nic_mseed_rechdr_t h; int32_t s[8];
        nic_mseed_read_record(out, (size_t)ns, &h, s, 8);
        check("sub-Hz rate round-trips", fabs(h.rate_hz - 0.1) < 1e-9 && h.nsamples == 5);
    }

    /* reclen = 4096 round-trips too */
    nic_mseed_params_t pbig = { "NQ", "ST01", "", "HHZ", 100.0, NIC_STEIM2, 4096 };
    long nb = nic_mseed_write_stream(&pbig, VEC_B, NB, 1700000000, 0.25, 1, out, CAP);
    check("reclen=4096 whole records", nb > 0 && nb % 4096 == 0);
    if (nb > 0) {
        got = 0; ok = 1;
        for (long off = 0; off < nb; off += 4096) {
            nic_mseed_rechdr_t h; int32_t s[1024];
            if (nic_mseed_read_record(out + off, (size_t)(nb - off), &h, s, 1024) != 0) { ok = 0; break; }
            for (uint16_t k = 0; k < h.nsamples && got < 2048; k++) rec_samp[got++] = s[k];
        }
        int eq = ok && got == NB;
        for (size_t k = 0; eq && k < NB; k++) if (rec_samp[k] != VEC_B[k]) eq = 0;
        check("reclen=4096 round-trip samples", eq);
    }

    printf("\nResult: %d/%d PASS | %d FAIL\n", passed, passed + failed, failed);
    return failed ? 1 : 0;
}
